### ISSUE TITLE
Increasing the client worker threads does not have any impact on parallel job executions

### ISSUE THREAD

**Issue by shahamit** [camunda/zeebe#14662]
**Describe the bug**

To increase the parallelism across multiple jobs we tried increasing the `zeebe.client.worker.threads` but this didn't have any impact on the execution. If there are multiple jobs running at the same time, they get stuck if one of them takes more than few secs.

**To Reproduce**

1. Deploy a bpmn file that consists of a job that takes more than 20 secs
2. Upload and execute 2 instances of this workflow. 

**Expected behaviour**

Both the instances should run in parallel after increasing the worker threads.

**Description**

On debugging the source I realised that we need to register multiple job workers for every job type then only increasing the worker threads take into effect. Probably that's because there is a job poller per job worker. 

**Environment:**
- OS: Linux
- Zeebe Version: 8.2.12
- Configuration: [spring-zeebe](https://github.com/camunda-community-hub/spring-zeebe): 8.2.3,

**Comment by nthed** [comment]
Tested above also with spring zeebe 8.2.4 and zeebe version 8.2.8.

**Comment by npepinpe** [comment]
ZDP triage: this sounds like purely like a client issue - let us know if this is not the case.

**Comment by shahamit** [comment]
@npepinpe - I am not sure but yes this could probably be an issue with the zeebe client. Since the [java client code](https://github.com/camunda/zeebe/tree/main/clients/java) is part of this repo, shouldn't it be tracked here for the fix? If this issue needs to be logged somewhere else kindly let me know.

**Comment by npepinpe** [comment]
No, this is the right repository, nothing to do on your side. We have two teams working on the same repo with slightly different focus, this was just a comment for them.

**Comment by remcowesterhoud** [comment]
ZPA Triage:
- Let's reproduce it and dig deeper
- Marking as mid severity as there is a noticeable impact, but there is a workaround
- @nthed Do you have any example code to reproduce it to make our lives easier?
- Marking it as `later`

**Comment by nthed** [comment]
Hi,

The initial issue was created by @shahamit , but I think we are experiencing the same behavior. 
It is also discussed within https://github.com/camunda-community-hub/spring-zeebe/issues/424 & https://github.com/camunda-community-hub/spring-zeebe/issues/493
Within the above issues, it is stated that in order for zeebe.client.worker.threads setting to take effect, the micrometer library has to be added to the project classpath otherwise the worker.threads setting has no effect.
When the micrometer library is added, the following behavior is experienced:

- It seems that the worker polls for jobs at t0. It fetches the jobs active (which may be less than the maximum number of threads) at that time and tries to execute them.
  If new jobs are created at t0 + some miliseconds while the job worker is executing the first batch, these jobs will not be picked up by the job worker  (until the whole first batch is completed) even though there are idle threads that could potentially serve them.
The issue becomes worse if the service consumed by the job worker is very slow (e.g. 30 sec response time), since this will mean that new jobs will be picked up by the job worker every 30 sec.

To test it , I created a simple rest service like the below which includes a thread sleep time, which I consume from a job worker using feign clients.

 ```
 @RequestMapping(
      value = "/myRequest",
      method = RequestMethod.POST,
      consumes = "application/json",
      produces = "application/json")
  public ResponseEntity<String> myRequest(@RequestBody JsonNode requestBody) {
log.info(requestBody.toString());
    try {
		Thread.sleep(30000);
	} catch (InterruptedException e) {
		// TODO Auto-generated catch block
		e.printStackTrace();
	}
    return new ResponseEntity<String>(HttpStatus.OK);
  }
```

 In the job worker I have added in my pom the following micrometer dependency, 

```
		<dependency>
		    <groupId>io.micrometer</groupId>
		    <artifactId>micrometer-registry-atlas</artifactId>
		</dependency>
```

In the application properties.file i have the following property configured

`zeebe.client.worker.threads = 10`

If the above is not sufficient I will try to create a stripped down version of the project for which I am experiencing the issue.

**Comment by micudaj** [comment]
Is there any progress made on this? We experience the same behaviour in our connector implmentation.

**Comment by npepinpe** [comment]
Are any of you on 8.4? Could you try using [job streaming](https://stage.docs.camunda.io/docs/next/apis-tools/java-client/job-worker/#job-streaming) instead? 

With this, the server will activate and push jobs out to your workers as they become available (grossly simplified). The `maxActiveJobs` on the client side becomes the maximum queue length of unhandled jobs. When a job is received, it comes one slot of the queue until the worker's job handler returns. So even while the job handler is working, the job is still considered as "queued", to avoid resource exhaustion. When the queue is full, jobs are yielded back (first rerouted to other available workers, but if none, they're made activate-able again in the engine).

So in your case, if you have 10 threads, and `maxActiveJobs == 100` it means you will be working on 10 jobs at most concurrently. The server will keep pushing up to 90 jobs more until the total is 100. When the job handler returns, then you will be able to get one more job.

**Comment by amitonlentra** [comment]
@npepinpe - we tried with 8.4.5. Increasing the `zeebe.client.worker.threads` did help after enabling `job streaming`. Few follow up questions
1. `maxJobsActive` is a configuration that applies to all threads of that worker or for each thread? Our workflow executions completed faster after we enabled `job streaming` but changing `maxJobsActive` from default (32) to 100 didn't show a visible difference, so I am assuming this is a per thread setting. 
2. Configuring `zeebe.client.worker.threads` to for e.g. 100 would mean each worker would have 100 threads and hence 100 jobs can be executed concurrently by each worker. We have 300 workers so this would mean 300 active workers if there 100+ jobs to be executed for each worker type, right?
3. On the zeebe client, do we need to tune any other property apart from these 3 to enable faster workflow executions? We are targeting 200-500 workflows/sec.

Thanks.
